File: app/src/main/java/ua/leonidius/rtlnotepad/dialogs/CloseTabDialog.java
package ua.leonidius.rtlnotepad.dialogs;

import android.app.Activity;
import android.app.AlertDialog;
import android.app.Dialog;
import android.app.DialogFragment;
import android.content.DialogInterface;
import android.os.Bundle;
import ua.leonidius.rtlnotepad.R;

/**
	This dialog is showed when user tries to close a tab with unsaved changes.
**/

public class CloseTabDialog extends DialogFragment implements AlertDialog.OnClickListener
{
	private Callback callback;
	private Activity activity;

	public CloseTabDialog(Activity activity) {
		super();
		this.activity = activity;
	}

	@Override
	public Dialog onCreateDialog(Bundle savedInstanceState)
	{
		AlertDialog.Builder adb = new AlertDialog.Builder(activity);
		adb.setMessage(R.string.close_tab_warning);
		adb.setNegativeButton(R.string.no, this);
		adb.setPositiveButton(R.string.yes, this);
		adb.setNeutralButton(R.string.cancel, this);
		return adb.create();
	}

	@Override
	public void onClick(DialogInterface p1, int id)
	{
		switch (id) {
			case Dialog.BUTTON_NEGATIVE:
				callback.callback(false);
				break;
			case Dialog.BUTTON_POSITIVE:
				callback.callback(true);
				break;
		}
	}
	
	public void setCallback(Callback callback) {
		this.callback = callback;
	}

	public interface Callback {
		void callback(boolean save);
	}

}

File: app/src/main/java/ua/leonidius/rtlnotepad/dialogs/CloseTabDialog.kt
package ua.leonidius.rtlnotepad.dialogs

import android.app.AlertDialog
import android.app.Dialog
import android.content.DialogInterface
import android.os.Bundle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import ua.leonidius.rtlnotepad.R

/**
 * This dialog is showed when user tries to close a tab with unsaved changes.
 */

class CloseTabDialog : BaseDialog(), DialogInterface.OnClickListener {

    private lateinit var viewModel: ViewModel

    companion object {
        fun create(callback : (Boolean) -> Unit) : CloseTabDialog {
            val dialog = CloseTabDialog()
            dialog.initializerFunction = {
                dialog.getViewModel().callback = callback
            }
            return dialog
        }
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val adb = AlertDialog.Builder(activity)
        adb.setMessage(R.string.close_tab_warning)
        adb.setNegativeButton(R.string.no, this)
        adb.setPositiveButton(R.string.yes, this)
        adb.setNeutralButton(R.string.cancel, this)
        return adb.create()
    }

    override fun onClick(p1: DialogInterface, id: Int) {
        when (id) {
            Dialog.BUTTON_NEGATIVE -> getViewModel().callback(false)
            Dialog.BUTTON_POSITIVE -> getViewModel().callback(true)
        }
    }

    private fun getViewModel() : Model {
        if (!this::viewModel.isInitialized) {
            viewModel = ViewModelProvider(this).get(Model::class.java)
        }
        return viewModel as Model
    }

    class Model : ViewModel() {
        lateinit var callback : (Boolean) -> Unit
    }

}